lib/devicetrust/native/device_windows.go

```diff
@@ -0,0 +1,489 @@
+		OsUsername:            u.Username,
+		OsVersion:             osVersion,
+		OsBuild:               osBuildNumber,
+		SystemSerialNumber:    systemSerial,
+		BaseBoardSerialNumber: baseBoardSerial,
+		ReportedAssetTag:      reportedAssetTag,
+	}
+	log.WithField(
+		"device_collected_data", dcd,
+	).Debug("Device data collected.")
+	return dcd, nil
+}
+
+// getDeviceCredential will only return the credential ID on windows. The
+// other information is determined server-side.
+func getDeviceCredential() (*devicepb.DeviceCredential, error) {
+	akPath, err := setupDeviceStateDir(os.UserHomeDir)
+	if err != nil {
+		return nil, trace.Wrap(err, "setting up device state directory")
+	}
+	tpm, err := openTPM()
+	if err != nil {
+		return nil, trace.Wrap(err, "opening tpm")
+	}
+	defer tpm.Close()
+
+	// Attempt to load the AK from well-known location.
+	ak, err := loadAK(tpm, akPath)
+	if err != nil {
+		return nil, trace.Wrap(err, "loading ak")
+	}
+	defer ak.Close(tpm)
+
+	credentialID, err := credentialIDFromAK(ak)
+	if err != nil {
+		return nil, trace.Wrap(err, "determining credential id")
+	}
+
+	return &devicepb.DeviceCredential{
+		Id: credentialID,
+	}, nil
+}
+
+func solveTPMEnrollChallenge(
+	challenge *devicepb.TPMEnrollChallenge,
+) (*devicepb.TPMEnrollChallengeResponse, error) {
+	akPath, err := setupDeviceStateDir(os.UserHomeDir)
+	if err != nil {
+		return nil, trace.Wrap(err, "setting up device state directory")
+	}
+
+	tpm, err := openTPM()
+	if err != nil {
+		return nil, trace.Wrap(err, "opening tpm")
+	}
+	defer tpm.Close()
+
+	// Attempt to load the AK from well-known location.
+	ak, err := loadAK(tpm, akPath)
+	if err != nil {
+		return nil, trace.Wrap(err, "loading ak")
+	}
+	defer ak.Close(tpm)
+
+	// Next perform a platform attestation using the AK.
+	log.Debug("Performing platform attestation.")
+	platformsParams, err := tpm.AttestPlatform(
+		ak,
+		challenge.AttestationNonce,
+		&attest.PlatformAttestConfig{
+			// EventLog == nil indicates that the `attest` package is
+			// responsible for providing the eventlog.
+			EventLog: nil,
+		},
+	)
+	if err != nil {
+		return nil, trace.Wrap(err, "attesting platform")
+	}
+
+	// First perform the credential activation challenge provided by the
+	// auth server.
+	log.Debug("Activating credential.")
+	activationSolution, err := ak.ActivateCredential(
+		tpm,
+		devicetrust.EncryptedCredentialFromProto(challenge.EncryptedCredential),
+	)
+	if err != nil {
+		return nil, trace.Wrap(err, "activating credential with challenge")
+	}
+
+	log.Debug("TPM enrollment challenge completed.")
+	return &devicepb.TPMEnrollChallengeResponse{
+		Solution: activationSolution,
+		PlatformParameters: devicetrust.PlatformParametersToProto(
+			platformsParams,
+		),
+	}, nil
+}
+
+// signChallenge is not implemented on windows as TPM platform attestation
+// is used instead.
+func signChallenge(_ []byte) (sig []byte, err error) {
+	return nil, devicetrust.ErrPlatformNotSupported
```

Comment:
Same as similar comment in device_darwin.go, use trace.BadParameter instead of ErrPlatformNotSupported.

Reply:
56bf50f84b